File: scripts/generate_index.js
const fs = require('fs');
const path = require('path');

function generate() {
    const basePath = path.join(__dirname, '..');
    const readmePath = path.join(basePath, 'README.md');
    const readmeContents = fs.readFileSync(readmePath, { encoding: 'utf-8' });
    const contents = [];
    const years = fs.readdirSync(basePath).filter((name) => /^\d+$/.test(name));
    years.sort((a, b) => +b - +a);
    for (const year of years) {
        contents.push(`<details><summary>${year}</summary>\n<p>\n`);
        const days = fs
            .readdirSync(path.join(basePath, year))
            .filter((name) => /^day\d+$/.test(name))
            .map((dayFolder) => ({ dayFolder, day: parseInt(dayFolder.slice(3)) }));
        days.sort((a, b) => a.day - b.day);
        for (const { dayFolder, day } of days) {
            contents.push(` - [Day ${day}](./${year}/${dayFolder}/src/main.rs)`);
        }
        contents.push(`\n</p>\n</details>`);
    }
    const newContents = readmeContents.replace(
        /<!-- INDEX-START -->.*?<!-- INDEX-END -->/s,
        `<!-- INDEX-START -->\n${contents.join('\n')}\n<!-- INDEX-END -->`
    );
    fs.writeFileSync(readmePath, newContents);
}

generate();

File: scripts/generate_index.rs
#!/usr/bin/env run-cargo-script
use std::fs;
use std::io;
use std::path::Path;

fn list_dir<S: AsRef<Path>>(dir: S) -> Result<Vec<String>, io::Error> {
    let mut entries = vec![];
    for child in fs::read_dir(dir)? {
        if let Some(name) = child?.path().file_name() {
            entries.push(String::from(name.to_str().unwrap().clone()));
        }
    }
    Ok(entries)
}

fn main() -> Result<(), io::Error> {
    let mut contents = String::new();
    let mut years = vec![];
    for dir in list_dir(".")? {
        if let Ok(year) = dir.parse::<u32>() {
            years.push(year);
        }
    }
    years.sort();
    for year in years.iter().rev() {
        let mut days = vec![];
        for folder in list_dir(format!("./{}", year))? {
            if folder.starts_with("day") {
                if let Ok(day) = folder[3..].parse::<u32>() {
                    days.push(day);
                }
            }
        }
        days.sort();

        contents
            .push_str(format!("<details><summary>{year}</summary>\n<p>\n\n", year = year).as_str());
        for day in days.iter().rev() {
            contents.push_str(
                format!(
                    " - [Day {day}](./{year}/day{day:02}/src/main.rs)\n",
                    day = day,
                    year = year
                )
                .as_str(),
            );
        }
        contents.push_str("\n</p>\n</details>\n");
    }

    let readme_contents = fs::read_to_string("./README.md")?;
    let start = readme_contents.find("<!-- INDEX-START -->").unwrap();
    let end = readme_contents.find("<!-- INDEX-END -->").unwrap();
    
    let mut result = String::new();
    result.push_str(&readme_contents[0..start]);
    result.push_str(&"<!-- INDEX-START -->\n");
    result.push_str(contents.as_str());
    result.push_str(&readme_contents[end..]);

    fs::write("README.md", result)?;

    Ok(())
}
